pkg/catalogv2/git/config.go

```diff
@@ -0,0 +1,144 @@
+package git
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/rancher/rancher/pkg/settings"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type Options struct {
+	Credential        *corev1.Secret
+	CABundle          []byte
+	InsecureTLSVerify bool
+	Headers           map[string]string
+}
+
+type git struct {
+	URL               string            `json:"URL"`
+	Directory         string            `json:"Directory"`
+	username          string            `json:"username"`
+	password          string            `json:"password"`
+	agent             *agent.Agent      `json:"agent"`
+	caBundle          []byte            `json:"caBundle"`
+	insecureTLSVerify bool              `json:"insecureTLSVerify"`
+	secret            *corev1.Secret    `json:"secret"`
+	headers           map[string]string `json:"headers"`
+	knownHosts        []byte            `json:"knownHosts"`
+}
+
+// gitForRepo constructs and returns a new git object for the given repository.
+// It requires a secret for authentication, the namespace, name of the repository,
+// the gitURL, a flag indicating if TLS verification should be skipped, and a CA bundle for SSL.
+// If the Git URL uses the SSH protocol, it checks if the URL is a valid SSH URL.
+// If the Git URL uses HTTP or HTTPS, it parses and verifies the URL.
+// It then constructs a directory path for the git repository.
+// It also sets an X-Install-Uuid header if an installation UUID is available.
+// If a CA bundle is provided, it converts the CA bundle from DER to PEM format, since Git requires PEM format.
+// In this case, insecureSkipTLS is set to false since a CA bundle is provided for secure communication.
+// Finally, it returns a new git object configured with these settings,
+// or an error if any step in this process fails.
+func gitForRepo(secret *corev1.Secret, namespace, name, gitURL string, insecureSkipTLS bool, caBundle []byte) (*git, error) {
+	isGitSSH, err := isGitSSH(gitURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify the type of URL %s: %w", gitURL, err)
+	}
+	if !isGitSSH {
+		u, err := url.Parse(gitURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse URL %s: %w", gitURL, err)
+		}
+		if u.Scheme != "http" && u.Scheme != "https" {
+			return nil, fmt.Errorf("invalid git URL scheme %s, only http(s) and git supported", u.Scheme)
+		}
+	}
```

Comment:
Shouldn't this be a separated method? Something like `validateGitUrl`?

Reply:
Part of the old code.